tool/cmd/kitex/versions/dependencies.go

```diff
@@ -0,0 +1,266 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package versions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/cloudwego/kitex"
+	"github.com/cloudwego/kitex/tool/internal_pkg/log"
+)
+
+var dm *dependencyManager
+
+func init() {
+	dm = &dependencyManager{
+		dependencies: make(map[string]*Dependency),
+	}
+	RegisterDependencies([]Dependency{
+		// todo(DMwangnima): support reading from yaml or json files
+		{
+			RefPath:        "github.com/cloudwego/kitex",
+			MinimalVersion: "v0.9.0",
+		},
+	})
+}
+
+// RegisterDependencies registers the dependencies needed to be checked
+func RegisterDependencies(deps []Dependency) {
+	if err := dm.RegisterDependencies(deps); err != nil {
+		panic(err)
+	}
+}
+
+// CheckDependencies is responsible for checking the compatibility of dependencies
+// and return the checking results.
+// After getting the CheckResult res, you can check the error with `Err()`:
+//
+//	if err := res.Err(); err != nil {
+//	    switch {
+//	    case errors.Is(err, ErrDependencyNotFound):
+//	    case errors.Is(err, ErrDependencyVersionNotSemantic):
+//	    case errors.Is(err, ErrDependencyVersionNotCompatible):
+//	    }
+//
+// Then you can get the version in go.mod with `GoModVersion()` and retrieve the dependency
+// information with `Dependency()`.
+func CheckDependencies() []*CheckResult {
+	return dm.Check()
+}
+
+// DefaultCheckDependenciesAndProcess provided default processing procedure to parse
+// CheckResult and prompt users
+func DefaultCheckDependenciesAndProcess() {
+	var promptSb strings.Builder
+	var shouldExit bool
+	for _, res := range dm.Check() {
+		err := res.Err()
+		switch {
+		case err == nil:
+			continue
+		case errors.Is(err, ErrDependencyNotFound):
+			// ignore ErrDependencyNotFound
+		case errors.Is(err, ErrDependencyVersionNotSemantic):
+			promptSb.WriteString(fmt.Sprintf("# %s %s is not semantic version in go.mod\n", res.Dependency().RefPath, res.GoModVersion()))
+		case errors.Is(err, ErrDependencyVersionNotCompatible):
+			shouldExit = true
+			promptSb.WriteString(fmt.Sprintf("# Kitex Cmd Tool %s is not compatible with %s %s\n", kitex.Version, res.Dependency().RefPath, res.GoModVersion()))
+			promptSb.WriteString(fmt.Sprintf("# You can upgrade %s to the minimum compatible version %s\n", res.Dependency().RefPath, res.Dependency().MinimalVersion))
+			promptSb.WriteString(fmt.Sprintf("go install %s@%s\n", res.Dependency().RefPath, res.Dependency().MinimalVersion))
+			promptSb.WriteString(fmt.Sprintf("# Or downgrade Kitex Cmd Tool to %s\n", res.GoModVersion()))
+			promptSb.WriteString(fmt.Sprintf("go install github.com/cloudwego/kitex/tool/cmd/kitex@%s\n", res.GoModVersion()))
+		}
+	}
+	if promptSb.Len() != 0 {
+		log.Warn(promptSb.String())
+	}
+	if shouldExit {
+		os.Exit(1)
```

Comment:
This impl is not suitable for unit test. A better approach is to return an error, and the caller is resposible for exiting the process.